add JSON parsing and hardware serial support to RS232 gateway

mreijnde opened the pull request:
## Description:
I added some features to the RS232 gateway I like to share. My use-case is logging JSON data to MQTT generated by an other microprocessor through a serial connection to OMG running on a ESP8266 or ESP32. 

Improvements:
- support for hardware serial (instead of software serial) to support higher baud rates
- support parsing of JSON received on the serial link to separate data in MQTT (sub) topics based on the JSON keys (up to the given nesting level).

EXAMPLE:
retrieved data on serial link: "{temperature: {sens1: 22, sens2: 23}, humidity: {sens1: 80, sens2: 60}}"

output to MQTT ( RS232maxJSONlevel=1):
            ./RS232toMQTT/temperature  ==> "{sens1: 22, sens2: 23}"
            ./RS232toMQTT/humidity     ==> "{sens1: 80, sens2: 60}"

output to MQTT ( RS232maxJSONlevel=2 (or higher)):
            ./RS232toMQTT/temperature/sens1 ==> 22
            ./RS232toMQTT/temperature/sens2 ==> 23
            ./RS232toMQTT/humidity/sens1 ==> 80
            ./RS232toMQTT/humidity/sens2 ==> 60

As default the original behavior is enabled to send the full RS232 like as MQTT message. By setting the RS232toMQTTmode to 1, the new JSON parsing mode can be enabled.

I tested it on a ESP32 and a ESP8266. The original  'raw' mode compiles on a Arduino Uno, the JSON variant does not fit in flash. The Arduino Mega does not compile, like on the original branch. I removed reference to it in the update.

Please have a look , I am open to suggestions to improve.


## Checklist:
  - [x] The pull request is done against the latest development branch
  - [x] Only one feature/fix was added per PR and the code change compiles without warnings
  - [x] I accept the [DCO](https://github.com/1technophile/OpenMQTTGateway/blob/development/docs/participate/development.md#developer-certificate-of-origin).

mreijnde committed:
add JSON parse option to RS232 gateway

mreijnde committed:
add hardware UART support to RS232 gateway

mreijnde committed:
fix func, ifdef & define names in RS232 gateway

mreijnde committed:
fix include warning

mreijnde committed:
changed default RS232toMQTTmode to raw (instead of JSON)

mreijnde committed:
removed reference to Arduino Mega in RS232 gateway config

mreijnde committed:
fix check of output buffer length

mreijnde committed:
fix formatting

mreijnde committed:
Revert "fix include warning"

This reverts commit a0ec9a187bb9c78556ffb2ffeae11d1c880e77a3.

1technophile commented:
@DieKatzchen would it be possible to review this PR as you are the author of the first version ?

DasKatzchen commented:
Gave it a look over, only thing I would change is putting the SoftwareSerial #include inside the #ifndef, since we wouldn't be using it in that instance. The serial stuff all looks about right (actually, I learned a thing or two), and I can't comment on the JSON formatting.

1technophile commented:
@mreijnde Thanks for the PR, could you take into account the comment please

mreijnde committed:
move SoftwareSerial include

mreijnde committed:
Revert "removed reference to Arduino Mega in RS232 gateway config"

This reverts commit 8082c1ad9c4cc8080a25b27275cf1eeac73b0986.

mreijnde committed:
update RS232 gateway doc

mreijnde commented:
I moved the #include as requested.

Furthermore:
- I added short explanation to the RS232 document
- Reverted the commit that removed pin config for Arduino Mega. Compile error earlier was due a misconfiguration.

1technophile commented:
Thanks!
